File: 2022/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Input
    long num;
    do
    {
        num = get_long("Number: ");
    }
    while (num < 0);

    // Determine length
    int len = 0;
    long card = num; //Variable "card" used to retain inputted number
    for (len = 0; num != 0; num /= 10, len++);
    if (len < 13 || len == 14 || len > 16)
    {
        // Is not the lenght of a valid credit card number
        printf("EOF\n");
    }

    // Sum of digits multiplied by 2
    long k = 10;
    long sum1 = 0;
    long calc1; //for first half of computation of sum1
    for (long n = 100; n <= 10000000000000000; n = n * 100)
    {
        calc1 = ((card % n) / k) * 2;
        sum1 = sum1 + (calc1 / 10) + (calc1 % 10);
        k = k * 100;
    }

    // Sum of digits not multiplied by 2
    long m = 10;
    long l = 1;
    long sum2 = 0;
    do
    {
        sum2 = sum2 + ((card % m) / l);
        m = m * 100;
        l = l * 100;
    }
    while (m <= 1000000000000000);

    // Sum of the two sums
    long sum = sum1 + sum2;

    // Validation
    while (card >= 100)
    {
        card /= 10;
    }
    if ((card > 55 || card < 34) || (sum % 10 != 0))
    {
        //Inputted number do not have the starting number of any credit card company
        printf("INVALID\n");
    }
    else
    {
        if ((len == 15) && (card == 34  || card == 37))
        {
            // Number is from AMEX
            printf("AMEX\n");
        }
        else if (len == 13 || len == 16)
        {
            if (card >= 40 && card <= 49)
            {
                // Number is from VISA
                printf("VISA\n");
            }
            else if (card >= 51 && card <= 55)
            {
                // Number is from MASTERCARD
                printf("MASTERCARD\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

File: 2023/pset6/sentimental-credit/credit.py
from cs50 import get_int


def main():
    # Get credit card number and number length
    num, num_len = get_num()
    print(num)

    # Determine if card is valid and its first two digits
    is_valid, num_dig = luhn_alg(num)

    # Determine and print what type of card it is
    card = typ_card(num_len, num_dig, is_valid)
    print(card)


def not_valid():
    # Return message for invalid card
    return "INVALID"


def get_num():
    # Ask for credit card number
    num = get_int("Number: ")

    # Calculate length
    num_len = len(str(num))

    # Only accept numbers with a length of 13 to 16
    if num_len in range(13, 17):
        return num, num_len
    else:
        print(not_valid())
        raise SystemExit(1)


def luhn_alg(num):
    # Initialize variables
    dig = dig1 = dig2 = sum_evn = sum_2odd = 0

    # Loop through digits from right to left
    while num > 0:
        # Get the rightmost digit and update variables
        dig2 = dig1
        dig1 = num % 10

        # Double every other digit starting from the second to last
        if dig % 2 == 0:
            sum_evn += dig1
        else:
            # If the doubled digit is two digits, add the digits together
            sum_2odd += ((2 * dig1) // 10) + ((2 * dig1) % 10)

        # Move to the next digit
        num //= 10
        dig += 1

    # Combine the last two digits to get the first two digits of the card number
    num_dig = (dig1 * 10) + dig2

    # Check if the sum of the digits is divisible by 10
    is_valid = (sum_evn + sum_2odd) % 10 == 0

    # Return the validity and first two digits of the card number
    return is_valid, num_dig


def typ_card(num_len, num_dig, is_valid):
    # Determine type of card based on length, first two digits, and validity
    if not is_valid:
        return not_valid()

    if num_len == 15 and num_dig in [34, 37]:
        return "AMEX"

    if num_len in [13, 16]:
        if 40 <= num_dig <= 49:
            return "VISA"
        if 51 <= num_dig <= 55:
            return "MASTERCARD"

    # Invalid if unidentified
    return not_valid()


if __name__ == "__main__":
    main()